### DISCUSSION TITLE
BDBrowser 1.9.5.20231111 - Launch Error

### DISCUSSION THREAD

**Discussion by JustMetallich** [tsukasa/BdBrowser#32] [Problems and issues]
BDBrowser stopped working at all. No settings, no plugins, no themes.

These errors appear when launching Discord:

%c[BDBrowser]%c %c[AsarUpdater]%c color: #3E82E5; font-weight: 700; color: #396CB8 Failed to download BetterDiscord Error: ENOENT: no such file or directory, open 'AppData/BetterDiscord/data/betterdiscord.asar'

152  console[type](`%c[BDBrowser]%c %c[${module}]%c`, "color: #3E82E5; font-weight: 700;", "", "color: #396CB8", "", ...message);
168    this.#log("error", module, ...message);
670 common_logger__WEBPACK_IMPORTED_MODULE_3__/* ["default"].error */ .Z.error(LOGGER_SECTION, "Failed to download BetterDiscord", err);

Uncaught (in promise) Error: BdBrowser Error: Downloading betterdiscord.asar or writing into VFS failed!

6682 throw new Error("BdBrowser Error: Downloading betterdiscord.asar or writing into VFS failed!");

I tried the "Delete Asar" option, but it doesn't work.

**Comment by tsukasa** [comment]
Greetings,

unfortunately, it is impossible to derive any conclusions from your post, as it lacks vital information, such as a log from the early loading of the page showing the initialization of the VFS and what happens during the download of the ASAR or even information about what browser you are using.

Therefore, I cannot provide you with any assistance except for the notice that both manually deleting the betterdiscord.asar from the VFS and triggering it through the extension options works as expected.

**Reply by JustMetallich** [reply]
Hello, I am using Google Chrome  v122.0.6261.129. 

> log from the early loading of the page showing the initialization of the VFS and what happens during the download of the ASAR

Where can I look for how to look at this log? I am sadly not a pro in this.

So I tried clearing all cache and cookies and BDWeb worked as intended(and it obviosly cleared all my plugins and themes as well, but that's not important), but then after a day or so it stopped working as well. I can repeat this expierience and try to look at this log.

**Reply by tsukasa** [reply]
Greetings,

thank you for wanting to dig deeper into the issue! The easiest way is to open the developer console by pressing F12, navigating the Console tab and pressing F5 to reload the current Discord tab.

You should then see messages similar to this:
```
[BDBrowser] [VFS] Database connection established.
[BDBrowser] [VFS] Pre-caching data from IndexedDB...
[BDBrowser] [VFS] Memory cache populated, took 164.50ms. VFS is ready.
[BDBrowser] [Frontend] Reading betterdiscord.asar in the VFS...
[BDBrowser] [Frontend] Registering callback.
[...]
[BDBrowser] [Frontend] Loading BetterDiscord renderer...
```

I am primarily interested the BDBrowser VFS initialization log entries and whether errors crop up in there, as well as the BDBrowser Frontend messages as those handle the loading of the ASAR file and running the initialization code for BetterDiscord.

You can remove any log entry with GatewaySocket, Flux, AuthenticationStore, DatabaseManager, and Spotify, as those are Discord system messages that have nothing to do with BdBrowser and might contain sensitive information.

It is likely that something goes wrong during the VFS initialization, the download of the ASAR file, or the initialization, and I hope to see something that gives me a hint as to what the cause could be.

**Reply by JustMetallich** [reply]
```
frontend.js:152 [BDBrowser] [AsarUpdater] Failed to download BetterDiscord Error: ENOENT: no such file or directory, open 'AppData/BetterDiscord/data/betterdiscord.asar'
    at getVfsErrorObject (frontend.js:5371:26)
    at Object.writeFileSync (frontend.js:5803:15)
    at BdAsarUpdater.downloadBetterDiscordAsar (frontend.js:680:91)
    at async Object.checkAndDownloadBetterDiscordAsar (frontend.js:2021:30)
    at async frontend.js:6692:10
frontend.js:152 [BDBrowser] [Frontend] Asar update reports failure.
VM765 47470.91c2d89….js:1 
 GET https://discord.com/api/v9/users/@me/connections/spotify/sxk7ep94o8fd7bt7knj0ibnxx/access-token 400 (Bad Request)
frontend.js:6693 Uncaught (in promise) Error: BdBrowser Error: Downloading betterdiscord.asar or writing into VFS failed!
    at frontend.js:6693:15
```
This is what I get

**Reply by tsukasa** [reply]
There should be the aforementioned messages about the VFS initialization before the BDBrowser AsarUpdater runs. Could you provide those log entries as well?

**Reply by JustMetallich** [reply]
```
[BDBrowser] [VFS] Database connection established.
frontend.js:152 [BDBrowser] [VFS] Pre-caching data from IndexedDB...
frontend.js:152 [BDBrowser] [VFS] Memory cache populated, took 18.00ms. VFS is ready.
frontend.js:152 [BDBrowser] [Frontend] No BetterDiscord asar present in VFS, will try to download a copy...
frontend.js:152 [BDBrowser] [AsarUpdater] Checking for latest BetterDiscord version...
backend.js:201 [BDBrowser] [Backend] Saved extension options: {deleteBdRendererOnReload: false, disableBdPluginsOnReload: false, disableBdRenderer: false}
frontend.js:152 [BDBrowser] [AsarUpdater] Latest stable BetterDiscord version is 1.10.1.
frontend.js:152 [BDBrowser] [AsarUpdater] Downloading BetterDiscord v1.10.1 into VFS...
```

**Reply by tsukasa** [reply]
Thank you!

To strike some possible suspects like VFS corruption off the list, I would advise to create a VFS backup, format the entire VFS and start fresh:
- Open the developer tools by pressing F12 and switch to the console.
- Create a VFS backup:
  ```
  require("fs").exportVfsBackup();
  ```
  A json file containing the current contents of your VFS will be downloaded.
- Once the file is downloaded, you can format the VFS:
  ```
  require("fs").formatVfs(true);
  ```
  This will reset the virtual filesystem contents to its initial state.
- After running the "format" command, press F5 to reload the page.
- The VFS initialization should run and hopefully the download of the betterdiscord.asar file will work now.

If that does not work, you have several other options you can try as a workaround:
- The easiest is [downloading the betterdiscord.asar](https://github.com/BetterDiscord/BetterDiscord/releases/tag/v1.10.1) manually, placing it in the "bd" subdirectory of BdBrowser's extension directory and reloading the extension through Chrome's "Extension" page. This will force BdBrowser to use the ASAR file from your local file system but has the disadvantage that you will need to manually update betterdiscord.asar in case there is a BetterDiscord update, as BdBrowser does not have access to write to your real filesystem.
- A somewhat more difficult option is creating your own VFS backup container from the contents of a real filesystem structure and loading that backup to provide BdBrowser with a sort of "repair":
  - See ["Create a Backup containing a custom BetterDiscord asar File"](https://github.com/tsukasa/BdBrowser?tab=readme-ov-file#creating-backup-files) in the BdBrowser README.md.
  - After creating your own backup file with just the "data/betterdiscord.asar" in it, open up the developer tools by pressing F12, switch to the console and run a VFS import:
    ```
    require("fs").importVfsBackup();
    ```
    A file picker will open, select the backup you created yourself. You should see the files being written to the VFS in the console.
  - Now press F5 to reload the Discord tab.
